src/validation.h

```diff
@@ -101,7 +101,21 @@ struct BlockHasher
     // this used to call `GetCheapHash()` in uint256, which was later moved; the
     // cheap hash function simply calls ReadLE64() however, so the end result is
     // identical
-    size_t operator()(const uint256& hash) const { return ReadLE64(hash.begin()); }
+    mutable CBlockIndex mock;
+    BlockHasher() : mock() {};
+    size_t operator()(CBlockIndex* const& ptr) const { return ReadLE64(ptr->GetBlockHash().begin()); }
+    // Helper for querying by hash
+    // e.g., map.find(map.hash_function()(h))
+    CBlockIndex* operator()(const uint256& hash) {
```

Comment:
In commit "Refactor BlockMap to use an unordered_set instead of an unordered_map" (fdc2f15c6b821f8ac0936cd35272a37305801723)

I was confused by the `idx.hash_function()(hash)` stuff initially, and think it could be clearer if you gave this method a different name like `FakeBlockIndex` instead of `operator()`. `idx.hash_function().FakeBlockIndex(hash)` would give a clue that the method is returning a fake block index pointer, not taking the hash of a hash, and also avoid overloading the same operator in two ways that do different things conceptually.

Also I guess you are using a shared mock CBlockIndex instance rather than creating temporary CBlockIndex objects for performance reasons, but it it would be interesting to know if taking a more straightforward with temporary mock objects would actually hurt performance:

```c++
struct FakeBlockIndex : public CBlockIndex
{
    FakeBlockIndex(const uint256& hash) { m_hash_block = hash; }
    FakeBlockIndex* pointer() { return this; }
};
```

Since it would allow simplifying:

```diff
-        BlockMap::const_iterator  it = m_blockman.m_block_index.find(m_blockman.m_block_index.hash_function()(hashAssumeValid));
+        BlockMap::const_iterator  it = m_blockman.m_block_index.find(FakeBlockIndex(hashAssumeValid).pointer());
```

and:

```diff
 struct BlockHasher
 {
     // this used to call `GetCheapHash()` in uint256, which was later moved; the
     // cheap hash function simply calls ReadLE64() however, so the end result is
     // identical
-    mutable CBlockIndex mock;
-    BlockHasher() : mock() {};
     size_t operator()(CBlockIndex* const& ptr) const { return ReadLE64(ptr->GetBlockHash().begin()); }
-    // Helper for querying by hash
-    // e.g., map.find(map.hash_function()(h))
-    CBlockIndex* operator()(const uint256& hash) {
-        mock.m_hash_block = hash;
-        return &mock;
-    }
 };
```

Reply:
I think it would since CBlockIndex is a very big struct?

but maybe we can make CBlockIndex have an abstract base class with just m_block_hash and then inherit from that for both? but then you have overhead *everywhere*...

Reply:
> I think it would since CBlockIndex is a very big struct?
> 
> but maybe we can make CBlockIndex have an abstract base class with just m_block_hash and then inherit from that for both? but then you have overhead _everywhere_...

I would probably not do the abstract base thing, since the reason for my suggestion was to simplify code, and trying to overload find that way would seem to add more complexity. Current approach here does seems fine. If dropping the hasher method would be bad for performance, I'd just rename it from `operator()` to `FakeBlockIndex` or something for clarity.

Reply:
i probably won't have time to work on this until mid april FYI, but i can pick it up sometime then.

if you'd like to pick it off and rebase it and push it through i would review it + ack.

Reply:
Good to know. I'm working on some other CBlockIndex changes so might want to do this. No hurry though!